src/Test/Spec/Reporter/TeamCity.purs

```diff
@@ -38,55 +33,104 @@ escape = Regex.replace'
+testFinished = teamcity' "" "testFinished"
+
+testFinishedIn :: WithDuration -> String
+testFinishedIn d = teamcity' ("duration" := (show $ trunc d.duration)) "testFinished" d
+
+testFailed :: WithMessage -> String
+testFailed d = teamcity' ("message" := d.message) "testFailed" d
 
-testSuiteStarted :: Array (Tuple String String) -> String
-testSuiteStarted = teamcity "testSuiteStarted"
+type ServiceMessage x =
+  { name :: String
+  , nodeId :: String
+  , parentNodeId :: Maybe String
+  | x
+  }
 
-testSuiteFinished :: Array (Tuple String String) -> String
-testSuiteFinished = teamcity "testSuiteFinished"
+type WithMessage = ServiceMessage (message :: String)
+type WithDuration = ServiceMessage (duration :: Number)
 
-testStarted :: Array (Tuple String String) -> String
-testStarted = teamcity "testStarted"
+serviceMessage :: String -> Path -> ServiceMessage ()
+serviceMessage name path =
+  let
+    nodeId = idFromPath path
+    parentNodeId = parentSuite path
+      <#> _.path
+      <#> idFromPath
+  in
+    { name, nodeId, parentNodeId }
 
-testIgnored :: Array (Tuple String String) -> String
-testIgnored = teamcity "testIgnored"
+withDuration :: Number -> ServiceMessage () -> WithDuration
+withDuration duration { name, nodeId, parentNodeId } =
+  { name, nodeId, parentNodeId, duration }
 
-testFinished :: Array (Tuple String String) -> String
-testFinished = teamcity "testFinished"
+withMessage :: String -> ServiceMessage () -> WithMessage
+withMessage message { name, nodeId, parentNodeId } =
+  { name, nodeId, parentNodeId, message }
 
-testFailed :: Array (Tuple String String) -> String
-testFailed = teamcity "testFailed"
+idFromPath :: Path -> String
+idFromPath path = path
+  <#> unwrap
+  <#> (\{ index, name } -> show index <> ":" <> fromMaybe "" name)
+  # Array.intercalate ","
 
 teamcityReporter :: Reporter
-teamcityReporter = defaultReporter "" case _ of
-  Event.Suite _ _ name -> do
-    put name
-    tellLn $ testSuiteStarted [ "name" /\ name ]
+teamcityReporter = defaultReporter Map.empty case _ of
+  Event.Suite _ path name -> do
+    void $ modify $ Map.insert path name
+    tellLn $ testSuiteStarted $ serviceMessage name path
   Event.SuiteEnd path -> do
-    name <- get
-    put $ nameFromPath path
-    tellLn $ testSuiteFinished [ "name" /\ name ]
-  Event.Test _ _ name ->
-    tellLn $ testStarted [ "name" /\ name ]
-  Event.Pending _ name -> do
-    tellLn $ testStarted [ "name" /\ name ]
-    tellLn $ testIgnored [ "name" /\ name ]
-    tellLn $ testFinished [ "name" /\ name ]
-  Event.TestEnd _ name (Success _ (Milliseconds millies)) -> do
-    tellLn $ testFinished
-      [ "name" /\ name
-      , "duration" /\ show (trunc millies)
-      ]
-  Event.TestEnd _ name (Failure error) -> do
-    tellLn $ testFailed
-      [ "name" /\ name
-      , "message" /\ show error
-      ]
-    tellLn $ testFinished [ "name" /\ name ]
+    name <- get <#> Map.lookup path <#> Maybe.fromMaybe ""
```

Comment:
When would the name for the path be missing? I'm assuming that would indicate a malfunction in the test runner itself, right? In that case, is it really correct to report test end for an empty-named test suite?

Reply:
This is a great idea!

What would ve the correct response? How do I implement it well? is `fromJust` the way to go?

Reply:
Well, I would just skip reporting if there is no started suite with that path:

```
    mName <- get <#> Map.lookup path
    for_ mName \name ->
      tellLn $ testSuiteFinished $ serviceMessage name path
```

That seems most reasonable: after all, if the path isn't in the map, we must have never issued a "suite started" event.